### ISSUE TITLE
Issues with memory usage and query index times, SQL and ForestDB implementations, 1.2 and bleeding edge

### ISSUE THREAD

**Issue by tikimcfee** [couchbase/couchbase-lite-android#797]
Issues with query index times, SQL and ForestDB implementations, 1.2 and bleeding edge

Hey there guys. Lots of good data here to help explain our findings with current 1.2 release and our use case. With the latest version for Android, one of our biggest issues with sync replication is starting to be curbed. While still running approximately 25-40 minutes depending on device for 100K documents, the replication batcher rework now allows an initial one-shot replication to finish - our issue of continuous mode reporting 'Idle' when all documents are not pulled still remains, so we must switch from one-shot to continuous mode after initial pull.

What we are founding now is an *increase in view index time, and massively increased memory footprint during initial replication and indexing*, with the potential to OOM on mid to low end devices. Attached to this issue are a number of snapshots indicating large allocation chunks in very small periods of time, and a rough stack trace of the area they are occuring in. I'm all too happy to add additional documentation for those as well should you want it.

Below is a snapshot of an emulated device (128mb heap, laptop based M.2 storage) running a full syncronization test from our test DB:

```
Sync Start:
02-26 15:09:20.858 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[0/1]  Error: (null)

	Checkpoint
02-26 15:20:00.281 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[52705/52905]  Error: (null)
02-26 15:20:01.079 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[52797/52997]  Error: (null)

	Checkpoint
02-26 15:30:18.501 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[95644/95844]  Error: (null)
02-26 15:30:19.165 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[95732/95932]  Error: (null)

	Sync Complete
02-26 15:33:47.966 V/Sync: [==> DOWN]: Continuous:[false] Active:[false]  LifeCycle:[Initial Synchronization]  Status:[Stopped]  Changes:[109340/109340]  Error: (null)
02-26 15:33:48.051 V/Sync: [UP ==>] : Continuous:[true] | Active:[true] | LifeCycle:[Current] | Status:[Active] | Changes:[0/0] | Error: (null) 	// Push replicator starts here
02-26 15:33:48.053 V/Sync: [==> DOWN]: Continuous:[true] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[0/0]  Error: (null)

02-26 15:33:54.738 D/DatabaseInit: Queries not yet initialized. Establishing views before first use. 
 - 02-26 15:33:54.738 D/Database: Preloading 'query1' 		3min
 - 02-26 15:36:51.792 D/Database: Preloading 'query2'		3min
 - 02-26 15:39:45.542 D/Database: Preloading 'query3'		4min
 - 02-26 15:44:01.579 D/Database: Preloading 'query4'		3min
 - 02-26 15:46:56.461 D/Database: Preloading 'query5'		3min
 - 02-26 15:49:52.677 D/Database: Preloading 'query6'		3min
 - 02-26 15:52:49.910 D/Database: Preloading 'query7'		3min
 - 02-26 15:55:45.910 D/Database: Preloading 'query8'		3min 
 - 02-26 15:58:41.263 D/Database: Preloading 'query9'		3min
 - 02-26 16:01:37.771 D/Database: Preloading 'query10'		3min
 - 02-26 16:05:46.373 D/Database: Preloading 'query11'		4min
 - 02-26 16:08:43.118 D/Database: Preloading 'query12'		3min

 	~~ This point and on, our switch from One-Shot to Continuous seemed to lock up on SQL database connections for a long time ~~
 ```

 More, this is a Motorola Razr M performing the same synchronization and indexing; it unfortunately could not finish indexing at all through a single run, as it continuously got caught on the SQLite connection pool bug. The app had to be reopened multiple times for the DB to be in a usable state:

 ```
02-29 12:35:26.949 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[0/0]  Error: (null)

02-29 12:39:45.411 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[12022/12228]  Error: (null)
02-29 12:39:46.736 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[12091/12321]  Error: (null)

02-29 12:43:17.752 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[22550/22730]  Error: (null)
02-29 12:43:17.753 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[22550/22750]  Error: (null)

02-29 12:49:55.588 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[41066/41266]  Error: (null)
02-29 12:49:57.263 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[41109/41309]  Error: (null)

02-29 12:50:49.413 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[43411/43611]  Error: (null)
02-29 12:50:51.733 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[43558/43727]  Error: (null)

02-29 12:56:39.715 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[59548/59748]  Error: (null)
02-29 12:56:41.847 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[59597/59797]  Error: (null)

02-29 12:59:11.074 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[66430/66630]  Error: (null)
02-29 12:59:12.735 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[66530/66730]  Error: (null)

02-29 13:02:14.797 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[74930/75130]  Error: (null)
02-29 13:02:16.846 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[75030/75230]  Error: (null)

02-29 13:05:46.081 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[84488/84688]  Error: (null)
02-29 13:05:47.832 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[84588/84788]  Error: (null)

02-29 13:09:22.302 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[94188/94388]  Error: (null)
02-29 13:09:25.189 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[94288/94488]  Error: (null)

02-29 13:14:32.118 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[107980/108172]  Error: (null)
02-29 13:14:32.246 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[107980/108203]  Error: (null)
02-29 13:15:05.533 V/Sync: [==> DOWN]: Continuous:[true] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[0/0]  Error: (null)
02-29 13:15:05.535 V/Sync: [UP ==>] : Continuous:[true] | Active:[true] | LifeCycle:[Current] | Status:[Active] | Changes:[0/0] | Error: (null)

02-29 13:15:14.391 D/Database: Preloading 'query1'
02-29 13:15:49.825 W/database: SQLiteConnectionPool: The connection pool for database 'our/database/location/dbname.cblite2/db.sqlite3' has been unable to grant a connection to thread 2406 (CBLReplicationWorker) with flags 0x2 for 30.000002 seconds.
                               Connections: 1 active, 0 idle, 0 available.
02-29 13:16:19.826 W/database: SQLiteConnectionPool: The connection pool for database 'our/database/location/dbname.cblite2/db.sqlite3' has been unable to grant a connection to thread 2406 (CBLReplicationWorker) with flags 0x2 for 60.035004 seconds.
                               Connections: 0 active, 1 idle, 0 available.
02-29 13:16:49.826 W/database: SQLiteConnectionPool: The connection pool for database 'our/database/location/dbname.cblite2/db.sqlite3' has been unable to grant a connection to thread 2406 (CBLReplicationWorker) with flags 0x2 for 90.036 seconds.
                               Connections: 0 active, 1 idle, 0 available.
02-29 13:17:19.827 W/database: SQLiteConnectionPool: The connection pool for database 'our/database/location/dbname.cblite2/db.sqlite3' has been unable to grant a connection to thread 2406 (CBLReplicationWorker) with flags 0x2 for 120.036 seconds.
                               Connections: 0 active, 1 idle, 0 available.
02-29 13:17:49.834 W/database: SQLiteConnectionPool: The connection pool for database 'our/database/location/dbname.cblite2/db.sqlite3' has been unable to grant a connection to thread 2406 (CBLReplicationWorker) with flags 0x2 for 150.041 seconds.
                               Connections: 0 active, 1 idle, 0 available.
02-29 13:18:19.836 W/database: SQLiteConnectionPool: The connection pool for database 'our/database/location/dbname.cblite2/db.sqlite3' has been unable to grant a connection to thread 2406 (CBLReplicationWorker) with flags 0x2 for 180.046 seconds.
                               Connections: 0 active, 1 idle, 0 available.
02-29 13:18:49.837 W/database: SQLiteConnectionPool: The connection pool for database 'our/database/location/dbname.cblite2/db.sqlite3' has been unable to grant a connection to thread 2406 (CBLReplicationWorker) with flags 0x2 for 210.046 seconds.
                               Connections: 1 active, 0 idle, 0 available.
02-29 13:19:19.838 W/database: SQLiteConnectionPool: The connection pool for database 'our/database/location/dbname.cblite2/db.sqlite3' has been unable to grant a connection to thread 2406 (CBLReplicationWorker) with flags 0x2 for 240.04701 seconds.
                               Connections: 0 active, 1 idle, 0 available.
 ```

One more test device, a Samsung Galaxy S3, running 1.2 with a ForestDB implementation instead of SQLite; this takes almost 2 hours to synchronize and index:

```
02-29 10:04:44.117 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[0/1]  Error: (null)
02-29 10:04:44.117 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[0/2]  Error: (null)

02-29 10:09:07.468 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[20752/20895]  Error: (null)
02-29 10:09:07.478 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[20752/20902]  Error: (null)

02-29 10:13:50.740 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[44081/44232]  Error: (null)
02-29 10:13:50.800 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[44081/44281]  Error: (null)

02-29 10:16:30.180 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[57333/57525]  Error: (null)
02-29 10:16:30.180 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[57333/57526]  Error: (null)

02-29 10:18:29.577 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[67071/67303]  Error: (null)
02-29 10:18:30.438 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[67106/67306]  Error: (null)

02-29 10:21:42.633 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[81643/81842]  Error: (null)
02-29 10:21:42.633 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[81643/81843]  Error: (null)

02-29 10:23:15.933 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[89416/89616]  Error: (null)
02-29 10:23:16.834 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[89464/89664]  Error: (null)

02-29 10:24:43.787 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[96648/96849]  Error: (null)
02-29 10:24:44.908 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[96741/96941]  Error: (null)

02-29 10:25:59.377 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[102894/103117]  Error: (null)
02-29 10:25:59.848 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[102946/103146]  Error: (null)

02-29 10:27:21.745 V/Sync: [==> DOWN]: Continuous:[false] Active:[true]  LifeCycle:[Initial Synchronization]  Status:[Active]  Changes:[109346/109346]  Error: (null)
02-29 10:27:21.745 E/Batcher: com.couchbase.lite.support.Batcher@422a38b8: flushing all objects (wait=false)
02-29 10:27:21.915 V/Sync: [==> DOWN]: Continuous:[false] Active:[false]  LifeCycle:[Initial Synchronization]  Status:[Stopped]  Changes:[109346/109346]  Error: (null)

02-29 10:27:23.147 D/Database: Preloading 'query1'
02-29 10:41:24.104 D/Database: Preloading 'query2'
02-29 10:51:52.925 D/Database: Preloading 'query3'
02-29 10:59:56.151 D/Database: Preloading 'query4'
02-29 11:06:34.746 D/Database: Preloading 'query5'
02-29 11:13:12.771 D/Database: Preloading 'query6'
02-29 11:20:00.206 D/Database: Preloading 'query7'
02-29 11:27:00.925 D/Database: Preloading 'query8'
02-29 11:34:06.789 D/Database: Preloading 'query9'
02-29 11:41:33.806 D/Database: Preloading 'query10'
02-29 11:49:53.749 D/Database: Preloading 'query11'
02-29 11:56:53.287 D/Database: Preloading 'query12'
```

 We have a number of tests that follow this pattern; it takes roughly an hour on a high end device for each of our views to be fully usable in the application, and longer for mid to low end devices. A recent test on iOS showed these results for its 1.2 release (It takes an iOS device running on the SQL implementation less than 8 minutes to perform the equivalent workload):
 ```
Initial Sync time  	: 00:04:30 (All 100K documents)
Screen 1 Show time  : 00:01:15 (Initial index of all views required for Screen 1)
Screen 2 time 		: 00:01:15 (Same as above)
 ```

--------------------------------------------------------
--------------------------------------------------------

Below are the mentioned snapshots of memory usage and allocation patterns. There is a lot of call time centered around the Jackson XML library, and most of those are Body or JSON allocations. For some context, these allocation snapshots are all post-replication and are occuring serially. We currently run an empty key query on all of our views for before moving in to the application, so as to not block the UI while waiting for these operations to complete (hence the 'Preloading' query logs above).  Hopefully these images can help suss out optimizations in allocation patterns of documents during indexing.

<img width="953" alt="1 second allocation" src="https://cloud.githubusercontent.com/assets/3186605/13406655/ad756514-def2-11e5-8aff-6a12624a7a8c.png">

<img width="947" alt="half second allocation" src="https://cloud.githubusercontent.com/assets/3186605/13406657/ad75e37c-def2-11e5-8cda-dcf089fe1a79.png">

<img width="936" alt="post index spike" src="https://cloud.githubusercontent.com/assets/3186605/13406656/ad75a6c8-def2-11e5-93c8-771c1b45ddd2.png">

<img width="574" alt="query_cursor_allocations" src="https://cloud.githubusercontent.com/assets/3186605/13406658/ad765348-def2-11e5-8805-0c716405eb70.png">

--------------------------------------------------------
--------------------------------------------------------

To summarize:

Our issue is with memory usage and index time in both SQLite and ForestDB in CBLite 1.2 . We are looking for any more information about this, or solutions to cut this number down (index on replication? multiple indexes at once?)

As always, please let us know if there is any more information we can provide!

**Comment by hideki** [comment]
Hi @tikimcfee,

Thank you for your detailed information, and sorry for slow response to your post.

About pull replication, memory usage is depends on number of documents, size of documents, number of revisions per document, number of conflicts in document, number of attachments, and size of each attachments. Please try to minimize these? If applicable, please consider to do preload database instead of initial pull replication? 

About updateIndexes, CBL indexes multiple views at once if view prefix name is same. Please not use same prefix name. We haven't spent time to investigate indexing time memory usage. We will investigate it.

If you could provide sample app to reproduce the problem with data, it will help us a lot. Please consider. 

Thank you!
Hideki
